gammapy/obs/observation.py

```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
                 raise ValueError('Invalid selection type: {}'.format(selection_type))
 
         return obs_table
+
+
+def recover_units(array, as_units):
+    """Utility functin to recover units.
+
+    After some numpy operations, `~astropy.units.Quantity`-like objects
+    loose their units. This function shoul recover them.
+
+    Parameters
+    ----------
+    array : `~numpy.ndarray` or `~astropy.units.Quantity`-like
+        Array without units.
+    as_units : int, float or `~astropy.units.Quantity`-like
+        Structure to imitate the units.
+
+    Returns
+    -------
+    array : int, float or `~astropy.units.Quantity`-like
+        Array with units.
+    """
+    try:
+        return Quantity(np.array(array), as_units.unit)
+    except:
+        # return unmodified
+        return array
+
+
+class ObservationGroups(object):
+
+    """Observation groups.
+
+    Class to define observation groups useful for organizing observation
+    lists into groups of observations with similar properties. The
+    properties and their binning are specified via
+    `~gammapy.obs.ObservationGroupAxis` objects.
+
+    The class takes as input a list of `~gammapy.obs.ObservationGroupAxis`
+    objects and defines 1 group for each possible combination of the
+    bins defined in all axes. The groups are identified by a unique
```

Comment:
The most common term for what you're describing here is "cartesian product":
https://en.wikipedia.org/wiki/Cartesian_product
https://docs.python.org/3.6/library/itertools.html#itertools.product

Can you use that term and maybe even link to the Wikipedia page?